src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs

```diff
@@ -62,25 +62,51 @@ public override TypeSyntax VisitArrayType(IArrayTypeSymbol symbol)
             {
                 ThrowIfNameOnly();
 
-                var underlyingNonArrayType = symbol.ElementType;
-                while (underlyingNonArrayType.Kind == SymbolKind.ArrayType)
+                ITypeSymbol underlyingType = symbol;
+
+                while (underlyingType is IArrayTypeSymbol innerArray)
                 {
-                    underlyingNonArrayType = ((IArrayTypeSymbol)underlyingNonArrayType).ElementType;
+                    underlyingType = innerArray.GetElementTypeWithAnnotatedNullability();
+
+                    if (underlyingType.GetNullability() == NullableAnnotation.Annotated)
+                    {
+                        // If the inner array we just moved to is also nullable, then
```

Comment:
@dotnet/roslyn-compiler: where is the spec for how we are (or are not) reordering array ranks when there are nullable annotations? I played around with a bit and arrived at this but an actual spec would be wonderful.

Reply:
This is quite confusing indeed.

---
In reply to: [312684703](https://github.com/dotnet/roslyn/pull/37892#discussion_r312684703) [](ancestors = 312684703)

Reply:
@jasonmalinowski We need to make sure our code fix for CS8600 accounts for this